### PR TITLE
Memory optimization for read EDIF

### PR DESCRIPTION
**Pull request by clavin-xlnx**
This PR contains three optimization aimed at reducing the memory footprint of an `EDIFNetlist`, primarily large ones.  The three optimizations are:

1) In `EDIFNet`, `Map<String,EDIFPortInst> portInsts` has been converted to `List<EDIFPortInst> portInsts`.  This is because 60% of `EDIFNetlist` memory consumption is from `HashMap` instances in `EDIFNet` and `EDIFCellInst`, with 67% of those instances coming from `EDIFNet`.  
2) In `EDIFCellInst`, `Map<String,EDIFPortInst> portInsts` has been converted to `List<EDIFPortInst> portInsts`.  This is because 60% of `EDIFNetlist` memory consumption is from `HashMap` instances in `EDIFNet` and `EDIFCellInst`, with 30% of those instances coming from `EDIFCellInst`.  
3) In `EDIFName` the default map type is being changed from `LinkedHashMap` to `HashMap`.  The previous was being used to preserve insertion order, but this is not strictly necessary.

Overall, these three optimizations **reduces memory footprint ~70%**, with the latter one contributing ~10% of the reduction.  Unfortunately, this does come at a cost to a few APIs that will exhibit behavior of `O(n)` instead of `O(1)`.  One way to mitigate this is to sort the lists in `EDIFNet` and `EDIFCellInst` which can reduce the impact somewhat to `O(log n)`.  In the end, a bit more testing and work is needed to identify the true impact of these changes and create alternative use modes to avoid the impacted APIs.  The main impacted APIs are:

- `EDIFNet.getPortInst(String fullName)`
- `EDIFNet.removePortInst(EDIFPortInst portInst)`
- `EDIFNet.removePortInst(String portInstFullName)`
- `EDIFCellInst.getPortInst(String name)`
- `EDIFCellInst.removePortInst(EDIFPortInst epr)`

### CODE REVIEW

**Comment by eddieh-xlnx** [comment]
Amazing!! I think sorted O(log(N)) is the way to go. Do you have a feel for how that might affect read runtime?

**Comment by clavin-xlnx** [comment]
### Performance Improvements for this PR
These are the average of 3 runs on a Windows 10 machine, Intel Core i7-9850H @ 2.6GHz with 32GB RAM, SSD, loading a 1.6GB EDIF file
### Memory Footprint 
Code Version | JDK 8 | JDK 11 | JDK 17|
---------------|--------|--------|--------|
Before PR | 4172MB |  4004MB | 4008MB | 
After PR   |  2183MB |  2147MB | 2155MB| 
**Improvement** |  **1.91x** |  **1.86x** | **1.86x** | 

### Runtime 
Code Version | JDK 8 | JDK 11 | JDK 17|
---------------|--------|--------|--------|
Before PR | 90.2s |  47.2s | 42.1s | 
After PR   |  60.5s |  35.7s | 44.7s| 
**Improvement** |  **1.49x** |  **1.32x** | **0.94x** |

**Comment by clavin-xlnx** [comment]
This PR contains #340.

File: src/com/xilinx/rapidwright/edif/EDIFTools.java [hunk 1]
```diff
@@ -238,16 +236,13 @@ public static void connectPortBus(EDIFCell topCell, EDIFCellInst src, EDIFCellIn
...
 
-			EDIFPortInst outputPortInst = srcPortMap.get(outputPortName);
-			EDIFPortInst inputPortInst = snkPortMap.get(inputPortName);
+			EDIFPortInst outputPortInst = src.getPortInst(outputPortName);
+			EDIFPortInst inputPortInst = snk.getPortInst(inputPortName);
 			
 			@SuppressWarnings("unused")
 			EDIFPortInst outputPort = new EDIFPortInst(outputPortInst.getPort(),net,outputPortInst.getIndex(),src);
 			@SuppressWarnings("unused")
 			EDIFPortInst inputPort = new EDIFPortInst(inputPortInst.getPort(),net,inputPortInst.getIndex(),snk);
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFTools.java:245]: Why is this necessary?

File: src/com/xilinx/rapidwright/edif/EDIFNet.java [hunk 2]
```diff
@@ -131,12 +128,35 @@ public EDIFPortInst createPortInst(EDIFPort port, int index, EDIFCellInst cellIn
...
+	 * {@link #getPortInstList()}.      
+	 * @return A map of EDIFPortInst names ({@link EDIFPortInst#getName()} to the corresponding objects.
+	 */
+	public Map<String, EDIFPortInst> getPortInstMap(){
+	    if(portInsts == null) return Collections.emptyMap();
+	    HashMap<String, EDIFPortInst> map = new HashMap<>();
+	    for(EDIFPortInst e : getPortInstList()) {
+	        map.put(e.getFullName(), e);
+	    }
+	    return map;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFNet.java:143]: Is this worth keeping around? GH is telling me `EDIFNet` is the only user, and all it seems to do is build this map, do a single lookup, then throw the map away.

File: src/com/xilinx/rapidwright/edif/EDIFCellInst.java [hunk 3]
```diff
@@ -88,57 +88,66 @@ public EDIFName getViewref() {
...
+    
     /**
-     * This gets a map of all the port refs on the cell instance.  
-     * @return A map of port ref names to port ref objects.
+     * Creates a new map of all the EDIFPortInst objects stored on this EDIFCellInst.  The new map
+     * contains a copy of EDIFPortInsts available at the time of invocation as returned from 
+     * {@link #getPortInstList()}.      
+     * @return A map of EDIFPortInst names ({@link EDIFPortInst#getName()} to the corresponding objects.
      */
     public Map<String, EDIFPortInst> getPortInstMap(){
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFCellInst.java:98]: Is this method worth keeping around? Looks like you've eliminated all users of it.

File: src/com/xilinx/rapidwright/edif/EDIFTools.java (hunk 1 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFTools.java:245]: This code was part of a debug demo, just trying not to break it.

File: src/com/xilinx/rapidwright/edif/EDIFCellInst.java (hunk 3 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFCellInst.java:98]: We could deprecate it and then, remove in a future release?

File: src/com/xilinx/rapidwright/edif/EDIFNet.java [hunk 4]
```diff
@@ -131,12 +128,35 @@ public EDIFPortInst createPortInst(EDIFPort port, int index, EDIFCellInst cellIn
...
 	}
 	
-	public Map<String,EDIFPortInst> getPortInstMap(){
-		return portInsts == null ? Collections.emptyMap() : portInsts;
+	/**
+	 * Creates a new map of all the EDIFPortInst objects stored on this net.  The new map
+	 * contains a copy of EDIFPortInsts available at the time of invocation as returned from 
+	 * {@link #getPortInstList()}.      
+	 * @return A map of EDIFPortInst names ({@link EDIFPortInst#getName()} to the corresponding objects.
+	 */
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFNet.java:136]:
```suggestion
	 * {@link #getPortInstList()}.      
	 * @return A map of EDIFPortInst names ({@link EDIFPortInst#getName()} to the corresponding objects.
	 * @deprecated
	 */
```

File: src/com/xilinx/rapidwright/edif/EDIFNet.java (hunk 2 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFNet.java:143]: I think I meant to tag it as deprecated

File: src/com/xilinx/rapidwright/edif/EDIFPortInstList.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            int compare = left.getName().compareTo(rightInstName);
+            return compare == 0 ? (-1*rightPortInstName.length()) : compare;
+        } else if(rightInstName == null) {
+            int compare = left.getCellInst().getName().compareTo(rightPortInstName);
+            return compare == 0 ? left.getName().length() : compare;
+        }
+        int compare = left.getCellInst().getName().compareTo(rightInstName);
+        return compare == 0 ? left.getName().compareTo(rightPortInstName) : compare; 
+    }
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFPortInstList.java:86]: @eddieh-xlnx - take extra care looking at this code.

File: src/com/xilinx/rapidwright/edif/EDIFCellInst.java [hunk 6]
```diff
@@ -151,8 +161,20 @@ public EDIFPort getPort(String name){
...
     public Collection<EDIFPortInst> getPortInsts(){
-        return getPortInstMap().values();
+        return portInsts == null ? Collections.emptyList() : portInsts;
+    }
+    
+    /**
+     * Gets the sorted ArrayList of EDIFPortInsts on this cell instance.
+     * @return The sorted list of EDIFPortInst objects on this cell.
+     */
+    public EDIFPortInstList getPortInstList() {
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFCellInst.java:176]: Does this new method need to exist, since it appears to be identical to `getPortInsts()` above except it returns a specific `EDIFPortInstList` type rather than a `Collection<EDIFPortInst>`? The only new user I'm seeing is:
https://github.com/Xilinx/RapidWright/blob/cd2afa438ca611ba52e1c62a704fa8b5a26a273e/src/com/xilinx/rapidwright/edif/EDIFCellInst.java#L102
which iterates over the object adding each value into a `HashMap`?

File: src/com/xilinx/rapidwright/edif/EDIFNet.java [hunk 7]
```diff
@@ -131,12 +128,36 @@ public EDIFPortInst createPortInst(EDIFPort port, int index, EDIFCellInst cellIn
...
 	public Collection<EDIFPortInst> getPortInsts(){
-		return portInsts == null ? Collections.emptyList() : portInsts.values();
+		return portInsts == null ? Collections.emptyList() : portInsts;
+	}
+	
+	/**
+	 * Gets the sorted ArrayList of EDIFPortInsts on this net.
+	 * @return The sorted list of EDIFPortInst objects on this net.
+	 */
+	public EDIFPortInstList getPortInstList() {
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFNet.java:159]: Similarly, is this not identical to `getPortInsts()` above?

File: src/com/xilinx/rapidwright/edif/EDIFPortInstList.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+     * @return 0 if the left and corresponding right Strings are equal.  A number less than 0 if
+     * left is lexicographically before right, or a number greater than 0 if left is after right. 
+     */
+    private int compare(EDIFPortInst left, String rightInstName, String rightPortInstName) {
+        if(left.getCellInst() == null) {
+            if(rightInstName == null) {
+                return left.getName().compareTo(rightPortInstName);
+            }
+            int compare = left.getName().compareTo(rightInstName);
+            return compare == 0 ? (-1*rightPortInstName.length()) : compare;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFPortInstList.java:79]: Does `-(rightPortInstName.length())` not work here?

File: src/com/xilinx/rapidwright/edif/EDIFPortInstList.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+     * considered port instance to compare left against.  
+     * @param rightPortInstName This is the port instance name {@link EDIFPortInst#getName()} of the
+     * considered port instance to compare left against.
+     * @return 0 if the left and corresponding right Strings are equal.  A number less than 0 if
+     * left is lexicographically before right, or a number greater than 0 if left is after right. 
+     */
+    private int compare(EDIFPortInst left, String rightInstName, String rightPortInstName) {
+        if(left.getCellInst() == null) {
+            if(rightInstName == null) {
+                return left.getName().compareTo(rightPortInstName);
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFPortInstList.java:76]:
```suggestion
                // left and right are both a top-level port insts, compare their port insts name only
                return left.getName().compareTo(rightPortInstName);
```

File: src/com/xilinx/rapidwright/edif/EDIFPortInstList.java (hunk 8 above)
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFPortInstList.java:79]:
```suggestion
            // left is a top-level port inst, right is not. Compare left's port inst name with the right's inst name
            int compare = left.getName().compareTo(rightInstName);
            // If the two happen to be equal, then left's full name is the prefix of right's full name and so should go before, with the remaining length the number of characters that differ
            return compare == 0 ? (-1*rightPortInstName.length()) : compare;
```

Also, do you need to account for the separator between inst name and port inst name that would otherwise appear when doing `getFullName()`?
After thinking about it some more, I think not (because this situation only occurs when comparing a top-level port inst with a non-top-level inst) unless the return value being off-by-one is critically important to the performance of the sorting algorithm?

File: src/com/xilinx/rapidwright/edif/EDIFPortInstList.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            int compare = left.getName().compareTo(rightInstName);
+            return compare == 0 ? (-1*rightPortInstName.length()) : compare;
+        } else if(rightInstName == null) {
+            int compare = left.getCellInst().getName().compareTo(rightPortInstName);
+            return compare == 0 ? left.getName().length() : compare;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFPortInstList.java:82]:
```suggestion
            // right is a top-level port inst, but left is not. Compare left's inst name with right's port inst name.
            int compare = left.getCellInst().getName().compareTo(rightPortInstName);
            // If the two happen to be equal, then right's full name is a prefix of left's full name and thus left should go after right.
            return compare == 0 ? left.getName().length() : compare;
```

File: src/com/xilinx/rapidwright/edif/EDIFPortInstList.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        }
+        int compare = left.getCellInst().getName().compareTo(rightInstName);
+        return compare == 0 ? left.getName().compareTo(rightPortInstName) : compare;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFPortInstList.java:85]:
```suggestion
        // left and right are both non top-level port insts. Compare their inst names first, then compare their port inst names.
        int compare = left.getCellInst().getName().compareTo(rightInstName);
        return compare == 0 ? left.getName().compareTo(rightPortInstName) : compare; 
```

File: src/com/xilinx/rapidwright/edif/EDIFNet.java (hunk 4 above)
Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFNet.java]: Looks good.

File: src/com/xilinx/rapidwright/edif/EDIFCellInst.java [hunk 12]
```diff
@@ -88,57 +88,66 @@ public EDIFName getViewref() {
...
     }
-
+    
     /**
-     * This gets a map of all the port refs on the cell instance.  
-     * @return A map of port ref names to port ref objects.
+     * Creates a new map of all the EDIFPortInst objects stored on this EDIFCellInst.  The new map
+     * contains a copy of EDIFPortInsts available at the time of invocation as returned from 
+     * {@link #getPortInstList()}.      
+     * @return A map of EDIFPortInst names ({@link EDIFPortInst#getName()} to the corresponding objects.
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFCellInst.java:96]:
```suggestion
     * @return A map of EDIFPortInst names ({@link EDIFPortInst#getName()} to the corresponding objects.
     * @deprecated
```

**Review by eddieh-xlnx — APPROVED** [review]
Just a few minor changes from me.

File: src/com/xilinx/rapidwright/edif/EDIFCellInst.java (hunk 12 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFCellInst.java:96]: Is this change tagging it to be `@deprecated`?  It seems like it already is.

File: src/com/xilinx/rapidwright/edif/EDIFPortInstList.java (hunk 8 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFPortInstList.java:80]: No, you are correct.  Since we are handling the compare strictly outside of the `getFullName()` scenario, it never becomes an issue.

File: src/com/xilinx/rapidwright/edif/EDIFCellInst.java (hunk 12 above)
Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFCellInst.java:96]: My mistake, sorry!
